### PR TITLE
Modularise all the things

### PR DESCRIPTION
**Pull request by 40Cakes**

### CODE REVIEW

File: modules/FlaskServer.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+from modules.mmf.Trainer import GetTrainer
+
+# TODO there's an issue with relative pathing, the static folder can't be found
+
+def httpServer(): # Run Flask server to make bot data available via HTTP GET
+    try:
+        log = logging.getLogger('werkzeug')
+        log.setLevel(logging.ERROR)
+
+        server = Flask(__name__,static_folder="interface")
```
Comment (alexm96) [inline modules/FlaskServer.py:18]: Not sure if this would work, but something like "../interface"

File: modules/FlaskServer.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+        CORS(server)
+        @server.route('/dashboard',methods=['GET'])
+        def req_dashboard():
+            return flask.render_template("dashboard.html")
+        @server.route('/dashboard/pokedex',methods=['GET'])
+        def req_dashboard_pokedex():
+            return flask.render_template("pokedex.html")
+        @server.route('/trainer_data', methods=['GET'])
+        def req_trainer_data():
+            trainer = GetTrainer()
```
Comment (alexm96) [inline modules/FlaskServer.py:28]: Could probably do 
```
def get_and_return_xyz(func:Callable):
    if callable_result:=func():
        return jsonify(callable_result)
    abort(503)
```
and replace all of the duplicated code

**Comment by 40Cakes** [comment]
Added `ui_ip and ``ui_port` to config but `127.0.0.1:8888` is hard coded in the web templates atm

File: modules/mmf/Pokemon.py [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+import numpy
+import fastjsonschema
+from datetime import datetime
+from modules.Config import GetConfig
+from modules.mmf.Common import LoadJsonMmap
+from modules.mmf.Trainer import GetTrainer
+
+config = GetConfig()
+
+with open("./data/items.json", mode="r", encoding="utf-8") as f:
```
Comment (alexm96) [inline modules/mmf/Pokemon.py:11]: Could probably pull this out as a function taking file_path:str as an arg

File: modules/mmf/Pokemon.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+            pokemon["move_3_pp"] = pokemon["pp"][2]
+            pokemon["move_4_pp"] = pokemon["pp"][3]
+            pokemon["type_1"] = pokemon["type"][0]
+
+            pokemon["type_2"] = "None"
+            if len(pokemon["type"]) == 2:
+                pokemon["type_2"] = pokemon["type"][1]
+
+            pokemon["enrichedMoves"] = []
+            for move in pokemon["moves"]:
```
Comment (alexm96) [inline modules/mmf/Pokemon.py:156]: can use list comprehension
'''
pokemon["enrichedMoves"]=[move for move in pokemon["moves"]]
'''

File: modules/mmf/Pokemon.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            return placeholder_pokemon
+        return None
+    except Exception as e:
+        print(str(e))
+        return placeholder_pokemon
+
+def GetParty():
+    try:
+        party = LoadJsonMmap(8192, "bizhawk_party_data-" + config["bot_instance_id"])["party"]
+        if party:
```
Comment (alexm96) [inline modules/mmf/Pokemon.py:201]: same thing here for list comp:
```
party_list=[pokemon for pokemon in party if PokemonValidator(pokemon)]
```

File: modules/mmf/Common.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    # BizHawk writes game information to memory mapped files every few frames (see pokebot.lua)
+    # See https://tasvideos.org/Bizhawk/LuaFunctions (comm.mmfWrite)
+    try:
+        shmem = mmap.mmap(0, size, file)
+        if shmem:
+            bytes_io = io.BytesIO(shmem)
+            byte_str = bytes_io.read()
+            json_obj = json.loads(byte_str.decode("utf-8").split("\x00")[0]) # Only grab the data before \x00 null chars
+            return json_obj
+        else: return False
```
Comment (alexm96) [inline modules/mmf/Common.py:16]: would it be possible for return None here instead? Kind of strange to return either a json object, bool, or None 😅

**Comment by 40Cakes** [comment]
Getting this error every 30 mins or so
```
line 642, in log_encounter
    if last_opponent_personality != pokemon["personality"]:
                                    ~~~~~~~^^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not subscriptable
```

**Pull request merged**